interp/src/interpreter/interpret_group.rs

```diff
@@ -1,357 +1,479 @@
+    }
+
+    fn update_val(&mut self, port: &ir::Port, value: OutputValue) {
+        self.working_env.insert(port as *const ir::Port, value);
+    }
+
+    fn get_as_val(&self, port: &ir::Port) -> &Value {
+        match self.get(port) {
+            OutputValueRef::ImmediateValue(iv) => iv.get_val(),
+            OutputValueRef::LockedValue(tlv) => tlv.get_val(),
+            OutputValueRef::PulseValue(pv) => pv.get_val(),
+        }
+    }
+
+    fn do_tick(&mut self) -> Vec<*const ir::Port> {
+        self.backing_env.clk += 1;
+
+        let mut new_vals = vec![];
+        let mut w_env = std::mem::take(&mut self.working_env);
+
+        self.working_env = w_env
+            .drain()
+            .filter_map(|(port, val)| match val {
+                OutputValue::ImmediateValue(iv) => {
+                    self.backing_env.insert(port, iv);
+                    None
+                }
+                out @ OutputValue::PulseValue(_)
+                | out @ OutputValue::LockedValue(_) => {
+                    let old_val = out.get_val().clone();
+
+                    match out.do_tick() {
+                        OutputValue::ImmediateValue(iv) => {
+                            if iv != old_val {
+                                new_vals.push(port)
+                            }
+                            self.backing_env.insert(port, iv);
+                            None
+                        }
+                        v @ OutputValue::LockedValue(_) => Some((port, v)),
+                        OutputValue::PulseValue(pv) => {
+                            if *pv.get_val() != old_val {
+                                new_vals.push(port);
+                            }
+                            Some((port, pv.into()))
+                        }
+                    }
+                }
+            })
+            .collect();
+        new_vals
+    }
+
+    fn collapse_env(mut self) -> Environment {
+        let working_env = self.working_env;
+
+        for (port, v) in working_env {
+            match v {
+                OutputValue::ImmediateValue(iv) => {
+                    self.backing_env.insert(port, iv)
+                }
+                OutputValue::LockedValue(tlv) => {
+                    if tlv.unlockable() {
+                        let iv = tlv.unlock();
+                        self.backing_env.insert(port, iv);
+                    } else {
+                        panic!("Group is done with an invalid value?")
+                    }
+                }
+                OutputValue::PulseValue(v) => {
+                    self.backing_env.insert(port, v.take_val())
                 }
-                map.insert(cb.clone_name(), ports);
             }
-            _ => panic!("component"),
         }
+        self.backing_env
+    }
+}
+
+// possibly #[inline] here later? Compiler probably knows to do that already
+fn get_done_port(group: &ir::Group) -> RRC<ir::Port> {
```

Comment:
Make a decision about this comment. Either add the attribute or remove the comment. No need for a TODO here.